scipy/stats/stats.py

```diff
@@ -1037,9 +1039,58 @@ def moment(a, moment=1, axis=0):
        is None. The denominator for the moment calculation is the number of
        observations, no degrees of freedom correction is done.
 
+    See also
+    --------
+    kurtosis, skew, describe
+
+    Notes
+    -----
+    [1]_ The n-th moment of a real-valued continuous function f(x) of a real
+    variable about a value c is:
+    .. math::
+
+    m_k = \frac{1}{n} \sum_{i = 1}^n (x_i - \bar{x})^k
+
+    This function uses exponentiation by squares [2]_ for efficiency.
+
+    References
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Moment_(mathematics)
+
+    .. [2] http://eli.thegreenplace.net/2009/03/21/efficient-integer-exponentiation-algorithms
     """
     a, axis = _chk_asarray(a, axis)
-    if moment == 1:
+
+    if a.size == 0:
+        # empty array, return nan(s) with shape matching `moment`
+        if np.isscalar(moment):
+            return np.nan
+        else:
+            return np.ones(np.asarray(moment).shape, dtype=np.float64) * np.nan
+
+    # for array_like moment input, return a value for each.
+    if not np.isscalar(moment):
+        mmnt = [_moment(a, i, axis) for i in moment]
+        return np.array(mmnt)
+    else:
+        return _moment(a, moment, axis)
+
+def _moment(a, moment, axis):
+    if not issubclass(type(moment), int):
+        raise ValueError("All moment parameters must be integers")
```

Comment:
This won't work well:

    In [7]: issubclass(np.int32, int)
    Out[7]: False

If you want to do a type check, this will work:

    np.issubdtype(dt, np.integer)  # works for all Python and Numpy integer types

Another and maybe simpler way to do it that would also work for floats that are integer-valued (so 1.0 but not 1.2):

    if not np.abs(moment - np.round(moment)) > 0:
        # Accept ints and integer-valued floats